### DISCUSSION TITLE
OIDC: API only login

### DISCUSSION THREAD

**Discussion by oyatsuk** [coder/coder#13662] [Feature Request]
Hi,

I am planning to build my own API on top of this solution so I am interested in using only API calls without an UI. I configured OIDC and I am able to login using user interface. 

**Is it possible to authenticate via API calls?** 

I can see that to send authenticated request I need "Coder-Session-Token" as a header or query parameter but I not completely understand how to get that token via API using OICD authentication.

**Comment by Emyrk** [comment]
@Majardom the problem you will run into is the OIDC "allow" page.

The flow is:

1. Hit `https://deployment.coder.com/api/v2/users/oauth2/github/callback`. This returns a 307 redirect to the OIDC/oauth provider
2. Follow redirect to `https://github.com/login/oauth/authorize?...`
3. This is where you hit the "allow" page

![github-oauth2-prompt](https://github.com/user-attachments/assets/0ec38011-5043-4d25-b236-bc796d5eccde)

4. After allow, you get a 302 redirect back to `https://deployment.coder.com/api/v2/users/oauth2/github/callback?coder=....`. The code works, and you get a 307 to `/workspaces` with a cookie that contains the session token.


So a headless auth flow for OIDC can only work if the IDP can do so headless. Which probably will not work for initial auth flows, but you can sometimes do it with subsequent auth flows if everything has been approved before.

I think what would work best for you is not implemented yet, but described in this issue: https://github.com/coder/coder/issues/11901
